code/vocabulary.py

```diff
@@ -1,9 +1,66 @@
 from talon import Context, Module, actions, grammar
 from .user_settings import bind_list_to_csv, bind_word_map_to_csv
 
+mod = Module()
+ctx = Context()
+
+mod.list("vocabulary", desc="additional vocabulary words")
+mod.list("punctuation", desc="words for inserting punctuation into text")
+
+@mod.capture(rule="({user.vocabulary} | <word>)")
+def word(m) -> str:
+    try:
+        return m.vocabulary
+    except AttributeError:
+        return " ".join(actions.dictate.replace_words(actions.dictate.parse_words(m.word)))
+
+@mod.capture(rule="({user.vocabulary} | <phrase>)+")
+def text(m) -> str: return format_phrase(m)
+
+@mod.capture(rule="({user.vocabulary} | {user.punctuation} | <phrase>)+")
+def prose(m) -> str: return format_phrase(m)
+
+# TODO: unify this formatting code with the dictation formatting code, so that
+# user.prose behaves the same way as dictation mode.
+punctuation = set(".,-!?;:")
+no_space_after = set("/-(")
+def format_phrase(m):
+    words = capture_to_word_list(m)
+    result = ""
+    for i, word in enumerate(words):
+        if i > 0 and word not in punctuation and words[i - 1][-1] not in no_space_after:
+            result += " "
+        result += word
+    return result
+
+def capture_to_word_list(m):
+    words = []
+    for item in m:
+        words.extend(
+            actions.dictate.replace_words(actions.dictate.parse_words(item))
+            if isinstance(item, grammar.vm.Phrase) else
+            item.split(" "))
+    return words
+
+
+# ---------- LISTS (punctuation, additional/replacement words) ----------
+bind_list_to_csv(
+    "user.punctuation",
+    "punctuation.csv",
+    csv_headers=("Punctuation Mark", "Word or Phrase"),
+    default_values={
+        "period": ".",
+        "comma": ",",
+        "colon": ":",
+        "semicolon": ";", "semi colon": ";",
+        "question mark": "?",
+        "exclamation mark": "!",
+    }
+)
+
 
 # Default words that will need to be capitalized (particularly under w2l).
-_capitalize_defaults = [
+capitalize = [
```

Comment:
There are lots of variable name changes that aren't obviously an improvement and make this change harder to understand. Would be nice separate functional changes and stylistic changes, though not a show stopper.

Reply:
I've reverted the variable name changes.